File: E3_shiny_app/R/unified_search.R
#' Batch-capable, multi-field search across integrated E3 relations.

unified_search_exact_columns <- function() {
  c(
    "accession", "candidate_accession", "candidate_accessions",
    "member_accession", "parsed_accession", "primary_group_id", "group_id",
    "hog_id", "orthogroup_id", "orthofinder_orthogroup_ids", "cluster_id",
    "lead_cluster_id", "representative_id", "internal_id", "gene_id",
    "protein_id", "entry", "parsed_entry", "member_identifier",
    "raw_identifier", "matched_seed_ids_calculated",
    "discovery_matched_seed_ids_calculated", "candidate_accessions_for_cluster"
  )
}

unified_search_text_columns <- function() {
  c(
    "name", "protein_name", "gene_name", "seed_protein_names",
    "discovery_seed_protein_names", "identifier_value", "available_aliases",
    "description", "annotation", "organism", "species", "species_column"
  )
}

#' Parse pasted search terms safely and deterministically.
#'
#' @param value Pasted newline/comma/semicolon/tab-delimited text.
#' @param maximum_terms Maximum accepted unique terms.
#' @return Unique non-empty terms in input order.
parse_unified_search_terms <- function(value, maximum_terms = 200L) {
  maximum_terms <- as.integer(maximum_terms)
  if (length(maximum_terms) != 1L || is.na(maximum_terms) || maximum_terms < 1L) {
    stop("maximum_terms must be a positive integer.", call. = FALSE)
  }
  if (length(value) == 0L || is.na(value[[1L]])) return(character())
  terms <- unlist(strsplit(as.character(value[[1L]]), "[\r\n,;\t]+"))
  terms <- trimws(terms)
  terms <- terms[nzchar(terms)]
  terms <- terms[!duplicated(tolower(terms))]
  if (length(terms) > maximum_terms) {
    stop(
      paste0("At most ", maximum_terms, " unique search terms are accepted."),
      call. = FALSE
    )
  }
  terms
}

validate_unified_search_mode <- function(mode) {
  if (
    length(mode) != 1L ||
      is.na(mode) ||
      !mode %in% c("smart", "exact", "contains")
  ) {
    stop("Unsupported unified-search matching mode.", call. = FALSE)
  }
  mode
}

#' Build a one-call relation/column catalogue query.
#'
#' @param alias Attached database alias.
#' @return DuckDB SQL query.
build_unified_search_catalogue_query <- function(alias = "e3_resource") {
  safe_alias <- escape_sql_literal(sanitise_duckdb_alias(alias))
  paste0(
    "SELECT table_name AS relation_name, column_name FROM ",
    "information_schema.columns WHERE table_catalog = '", safe_alias,
    "' AND table_schema = 'main' ORDER BY table_name, ordinal_position"
  )
}

unified_search_condition <- function(column, term, mode, text_column) {
  field <- paste0("coalesce(CAST(source.", quote_duckdb_identifier(column),
    " AS VARCHAR), '')"
  )
  safe <- escape_sql_literal(tolower(term))
  contains <- paste0("instr(lower(", field, "), '", safe, "') > 0")
  exact <- paste0(
    "(lower(trim(", field, ")) = '", safe,
    "' OR list_contains(string_split(replace(lower(", field,
    "), '; ', ';'), ';'), '", safe, "'))"
  )
  if (mode == "contains" || (mode == "smart" && text_column)) {
    contains
  } else {
    exact
  }
}

#' Build a bounded query for one searchable relation.
#'
#' @param relation Relation name.
#' @param columns Source column names.
#' @param terms Parsed search terms.
#' @param mode Smart, exact or contains.
#' @param max_rows Maximum matching rows for this relation.
#' @param alias Attached database alias.
#' @return DuckDB SQL query or NULL when no fields are searchable.
build_unified_search_query <- function(
  relation,
  columns,
  terms,
  mode = "smart",
  max_rows = 250L,
  alias = "e3_resource"
) {
  mode <- validate_unified_search_mode(mode)
  terms <- parse_unified_search_terms(paste(terms, collapse = "\n"))
  if (length(terms) == 0L) stop("At least one search term is required.", call. = FALSE)
  max_rows <- max(1L, min(10000L, as.integer(max_rows)))
  exact_columns <- intersect(unified_search_exact_columns(), columns)
  text_columns <- intersect(unified_search_text_columns(), columns)
  searchable <- unique(c(exact_columns, text_columns))
  if (length(searchable) == 0L) return(NULL)
  source <- qualified_resource_relation(relation, alias)
  source_fields <- paste(vapply(columns, function(column) {
    paste0(
      "CAST(source.", quote_duckdb_identifier(column), " AS VARCHAR) AS ",
      quote_duckdb_identifier(column)
    )
  }, character(1L)), collapse = ", ")
  queries <- vapply(terms, function(term) {
    conditions <- vapply(searchable, function(column) {
      unified_search_condition(
        column,
        term,
        mode,
        column %in% text_columns
      )
    }, character(1L))
    matched_fields <- vapply(seq_along(searchable), function(index) {
      paste0(
        "CASE WHEN ", conditions[[index]], " THEN '",
        escape_sql_literal(searchable[[index]]), "' ELSE NULL END"
      )
    }, character(1L))
    paste0(
      "SELECT '", escape_sql_literal(term), "' AS _search_term, '",
      escape_sql_literal(relation), "' AS _relation, ",
      "concat_ws(';', ", paste(matched_fields, collapse = ", "),
      ") AS _matched_columns, ", source_fields, " FROM ", source,
      " source WHERE ", paste(conditions, collapse = " OR ")
    )
  }, character(1L))
  paste0(
    "SELECT * FROM (", paste(queries, collapse = " UNION ALL "),
    ") matches LIMIT ", max_rows
  )
}

#' Search every relation containing recognised fields.
#'
#' @param resource_source Flexible E3 resource source.
#' @param catalogue Relation/column catalogue.
#' @param terms Parsed search terms.
#' @param mode Matching mode.
#' @param max_rows_per_relation Per-relation row limit.
#' @param max_total_rows Total output limit.
#' @return Character-stable combined matching rows.
collect_unified_search_results <- function(
  resource_source,
  catalogue,
  terms,
  mode = "smart",
  max_rows_per_relation = 250L,
  max_total_rows = 10000L
) {
  max_total_rows <- max(1L, min(100000L, as.integer(max_total_rows)))
  relations <- unique(as.character(catalogue$relation_name))
  results <- list()
  for (relation in relations) {
    columns <- as.character(
      catalogue$column_name[catalogue$relation_name == relation]
    )
    query <- build_unified_search_query(
      relation,
      columns,
      terms,
      mode,
      max_rows_per_relation
    )
    if (is.null(query)) next
    rows <- tryCatch(
      collect_resource_query(resource_source, query),
      error = function(error) {
        warning(
          paste0("Search skipped relation ", relation, ": ", error$message),
          call. = FALSE
        )
        NULL
      }
    )
    if (!is.null(rows) && nrow(rows) > 0L) results[[relation]] <- rows
  }
  if (length(results) == 0L) {
    return(tibble::tibble(
      `_search_term` = character(),
      `_relation` = character(),
      `_matched_columns` = character()
    ))
  }
  combined <- dplyr::bind_rows(results)
  utils::head(combined, max_total_rows)
}

#' Summarise complete unified-search rows.
#'
#' @param matches Complete result rows.
#' @return Counts by term, relation and matched fields.
summarise_unified_search_results <- function(matches) {
  required <- c("_search_term", "_relation", "_matched_columns")
  missing <- setdiff(required, names(matches))
  if (length(missing) > 0L) {
    stop(
      paste("Search matches are missing:", paste(missing, collapse = ", ")),
      call. = FALSE
    )
  }
  if (nrow(matches) == 0L) {
    return(tibble::tibble(
      search_term = character(),
      relation = character(),
      matched_columns = character(),
      matching_rows = integer()
    ))
  }
  summary <- matches |>
    dplyr::count(
      .data$`_search_term`,
      .data$`_relation`,
      .data$`_matched_columns`,
      name = "matching_rows"
    )
  names(summary)[match(required, names(summary))] <- c(
    "search_term",
    "relation",
    "matched_columns"
  )
  dplyr::arrange(summary, .data$search_term, .data$relation, .data$matched_columns)
}

File: e3_python_app/src/e3app/unified_search.py
"""Defensive multi-item, multi-field search across loaded E3 relations."""

from __future__ import annotations

import logging
import re
from typing import Literal, Sequence

import pandas as pd

from e3app.data import list_relations, quote_identifier, relation_columns
from e3app.errors import AppError

LOGGER = logging.getLogger(__name__)

SearchMode = Literal["smart", "exact", "contains"]

EXACT_SEARCH_COLUMNS = {
    "accession",
    "entry",
    "entry_name",
    "evolutionary_group_key",
    "primary_group_id",
    "group_id",
    "hog_id",
    "orthogroup_id",
    "cluster_id",
    "lead_cluster_id",
    "representative_id",
    "representative_original_id",
    "protein_accession",
    "candidate_accession",
    "parsed_accession",
    "member_accession",
    "member_identifier",
    "raw_identifier",
    "parsed_entry",
    "gene_id",
    "identifier_value",
    "candidate_accessions",
    "candidate_accessions_for_cluster",
    "matched_seed_ids_calculated",
    "discovery_matched_seed_ids_calculated",
}

TEXT_SEARCH_COLUMNS = {
    "protein_name",
    "protein_names",
    "seed_protein_names",
    "discovery_seed_protein_names",
    "gene_name",
    "gene_names",
    "matched_gene_names",
    "identifier_value",
    "entry_name",
    "category",
    "seed_categories",
    "raw_header",
    "description",
}


def parse_search_terms(*, value: str, maximum_terms: int = 50) -> tuple[str, ...]:
    """Parse a pasted identifier/name list while preserving spaces within names.

    Args:
        value: Terms separated by newlines, commas, semicolons or tabs.
        maximum_terms: Defensive maximum number of unique terms.

    Returns:
        Unique, ordered search terms.

    Raises:
        AppError: If any term or the complete list exceeds safe bounds.
    """
    if not 1 <= maximum_terms <= 500:
        raise AppError("maximum_terms must be between 1 and 500")
    if not isinstance(value, str):
        raise AppError("search input must be text")
    pieces = re.split(r"[\n,;\t]+", value)
    terms: list[str] = []
    seen: set[str] = set()
    for piece in pieces:
        term = piece.strip()
        if not term:
            continue
        if len(term) > 200:
            raise AppError("each search term must contain at most 200 characters")
        key = term.casefold()
        if key not in seen:
            seen.add(key)
            terms.append(term)
    if len(terms) > maximum_terms:
        raise AppError(f"search accepts at most {maximum_terms} unique terms")
    return tuple(terms)


def validate_search_mode(*, mode: SearchMode) -> None:
    """Validate a unified-search matching mode."""
    if mode not in {"smart", "exact", "contains"}:
        raise AppError(f"Unsupported search mode: {mode}")


def searchable_columns(*, columns: Sequence[str]) -> tuple[list[str], list[str]]:
    """Classify exact/token and descriptive-text search fields.

    Args:
        columns: Relation column names.

    Returns:
        Exact/token fields and descriptive-text fields, in source order.
    """
    exact: list[str] = []
    text: list[str] = []
    for column in columns:
        lower = column.lower()
        if lower in EXACT_SEARCH_COLUMNS or re.search(
            r"(^|_)(hog|orthogroup|cluster|seed|accession|identifier|gene_id)(_|$)",
            lower,
        ):
            exact.append(column)
        if lower in TEXT_SEARCH_COLUMNS or re.search(
            r"(^|_)(protein|gene|seed).*name(s)?$|(^|_)(description|alias)(_|$)",
            lower,
        ):
            text.append(column)
    return exact, text


def _exact_condition(*, column: str, term_expression: str) -> str:
    """Build a case-insensitive scalar or semicolon-token match."""
    quoted = quote_identifier(column)
    value = f"upper(trim(coalesce(CAST({quoted} AS VARCHAR), '')))"
    term = f"upper(trim({term_expression}))"
    return (
        f"({value} = {term} OR "
        f"instr(';' || replace({value}, '; ', ';') || ';', "
        f"';' || {term} || ';') > 0)"
    )


def _contains_condition(*, column: str, term_expression: str) -> str:
    """Build a case-insensitive literal substring match."""
    quoted = quote_identifier(column)
    return (
        f"instr(lower(coalesce(CAST({quoted} AS VARCHAR), '')), "
        f"lower(trim({term_expression}))) > 0"
    )


def _relation_search_sql(
    *,
    relation: str,
    columns: Sequence[str],
    terms: Sequence[str],
    mode: SearchMode,
    maximum_rows: int,
) -> tuple[str, list[object]] | None:
    """Build one bound search query for a single relation."""
    exact_columns, text_columns = searchable_columns(columns=columns)
    if mode == "exact":
        selected = [(column, "exact") for column in exact_columns]
    elif mode == "contains":
        selected = [
            (column, "contains")
            for column in dict.fromkeys([*exact_columns, *text_columns])
        ]
    else:
        selected = [
            (
                column,
                "contains" if column in text_columns else "exact",
            )
            for column in dict.fromkeys([*exact_columns, *text_columns])
        ]
    if not selected:
        return None
    values = ", ".join("(?, ?)" for _ in terms)
    parameters: list[object] = []
    for order, term in enumerate(terms, start=1):
        parameters.extend((order, term))
    conditions: list[str] = []
    labels: list[str] = []
    for column, match_type in selected:
        condition = (
            _exact_condition(column=column, term_expression="t.search_term")
            if match_type == "exact"
            else _contains_condition(column=column, term_expression="t.search_term")
        )
        conditions.append(condition)
        labels.append(f"CASE WHEN {condition} THEN '{column}' END")
    sql = (
        f"WITH search_terms(search_order, search_term) AS (VALUES {values}) "
        f"SELECT t.search_order AS _search_order, t.search_term AS _search_term, "
        f"'{relation}' AS _relation, "
        f"concat_ws(';', {', '.join(labels)}) AS _matched_columns, source.* "
        f"FROM {quote_identifier(relation)} source CROSS JOIN search_terms t "
        f"WHERE {' OR '.join(conditions)} "
        "ORDER BY t.search_order "
        f"LIMIT {int(maximum_rows)}"
    )
    return sql, parameters


def collect_unified_search(
    *,
    connection: object,
    search_terms: Sequence[str],
    mode: SearchMode = "smart",
    maximum_rows_per_relation: int = 250,
    maximum_total_rows: int = 10_000,
) -> pd.DataFrame:
    """Search identifiers and names across every compatible loaded relation.

    Args:
        connection: Open read-only DuckDB connection.
        search_terms: One or several exact identifiers or name fragments.
        mode: Smart, exact/token or contains matching.
        maximum_rows_per_relation: Bound applied within each source relation.
        maximum_total_rows: Bound applied after combining relation results.

    Returns:
        Combined matches retaining relation, search-term and field provenance.
    """
    validate_search_mode(mode=mode)
    terms = tuple(dict.fromkeys(term.strip() for term in search_terms if term.strip()))
    if not terms:
        raise AppError("Enter at least one search term")
    if len(terms) > 50:
        raise AppError("search accepts at most 50 unique terms")
    if not 1 <= maximum_rows_per_relation <= 10_000:
        raise AppError("maximum_rows_per_relation must be between 1 and 10000")
    if not 1 <= maximum_total_rows <= 100_000:
        raise AppError("maximum_total_rows must be between 1 and 100000")
    frames: list[pd.DataFrame] = []
    for relation in list_relations(connection):
        columns = relation_columns(connection, relation)
        built = _relation_search_sql(
            relation=relation,
            columns=columns,
            terms=terms,
            mode=mode,
            maximum_rows=maximum_rows_per_relation,
        )
        if built is None:
            continue
        sql, parameters = built
        try:
            frame = connection.execute(sql, parameters).fetchdf()
        except Exception as exc:  # pragma: no cover - defensive per-relation isolation
            LOGGER.warning("Search skipped relation %s: %s", relation, exc)
            continue
        if not frame.empty:
            frames.append(frame)
    if not frames:
        return pd.DataFrame(
            columns=("_search_order", "_search_term", "_relation", "_matched_columns")
        )
    combined = pd.concat(frames, ignore_index=True, sort=False)
    combined = combined.sort_values(
        by=["_search_order", "_relation"],
        kind="stable",
    )
    return combined.head(maximum_total_rows).reset_index(drop=True)


def summarise_unified_search(*, matches: pd.DataFrame) -> pd.DataFrame:
    """Summarise match counts by pasted term and source relation."""
    required = {"_search_order", "_search_term", "_relation"}
    missing = sorted(required.difference(matches.columns))
    if missing:
        raise AppError("search results are missing columns: " + ", ".join(missing))
    if matches.empty:
        return pd.DataFrame(
            columns=("search_order", "search_term", "relation", "matched_rows")
        )
    summary = (
        matches.groupby(
            ["_search_order", "_search_term", "_relation"],
            dropna=False,
            sort=True,
        )
        .size()
        .rename("matched_rows")
        .reset_index()
        .rename(
            columns={
                "_search_order": "search_order",
                "_search_term": "search_term",
                "_relation": "relation",
            }
        )
    )
    return summary
